pkg/registry/core/service/ipallocator/controller/repair.go

```diff
@@ -216,64 +208,72 @@ func (c *Repair) runOnce() error {
 			// didn't need a cluster IP
 			continue
 		}
-		ip := net.ParseIP(svc.Spec.ClusterIP)
-		if ip == nil {
-			// cluster IP is corrupt
-			c.recorder.Eventf(&svc, v1.EventTypeWarning, "ClusterIPNotValid", "Cluster IP %s is not a valid IP; please recreate service", svc.Spec.ClusterIP)
-			runtime.HandleError(fmt.Errorf("the cluster IP %s for service %s/%s is not a valid IP; please recreate", svc.Spec.ClusterIP, svc.Name, svc.Namespace))
-			continue
-		}
 
-		// mark it as in-use
-		actualAlloc := c.selectAllocForIP(ip, rebuilt, secondaryRebuilt)
-		switch err := actualAlloc.Allocate(ip); err {
-		case nil:
-			actualStored := c.selectAllocForIP(ip, stored, secondaryStored)
-			if actualStored.Has(ip) {
-				// remove it from the old set, so we can find leaks
-				actualStored.Release(ip)
-			} else {
-				// cluster IP doesn't seem to be allocated
-				c.recorder.Eventf(&svc, v1.EventTypeWarning, "ClusterIPNotAllocated", "Cluster IP %s is not allocated; repairing", ip)
-				runtime.HandleError(fmt.Errorf("the cluster IP %s for service %s/%s is not allocated; repairing", ip, svc.Name, svc.Namespace))
+		for _, ip := range svc.Spec.ClusterIPs {
```

Comment:
Does this work during HA controlplane updates?  Is this controller guaranteed to get the updated API server?  Or do we have to handle the case of ClusterIPs being blank (and fall back on ClusterIP) ?

Reply:
I mean we had the ClusterIP=>ClusterIPs[0], then we removed as you requested. My understanding is it connects to self, no?